### PR TITLE
Multinode and Startup Fixes

### PR DESCRIPTION
**Pull request by jrhea**
## PR Description
- [x] added an artemis command to generate peer-ids
- [x] fixes to multi-node run that was broken after interop
- [x] Untangle the startup event mess

### Defer for another PR
- We need an LRU cache to filter duplicate gossip messages
- Splitting validators across nodes causes randao issues

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java [hunk 1]
```diff
@@ -145,10 +145,8 @@ public static BeaconStateWithCache initialize_beacon_state_from_eth1(
 
   public static BeaconStateWithCache initialize_beacon_state_from_eth1_new(
       Bytes32 eth1_block_hash, UnsignedLong eth1_timestamp, List<? extends Deposit> deposits) {
-    UnsignedLong genesis_time =
-        eth1_timestamp
-            .minus(eth1_timestamp.mod(UnsignedLong.valueOf(SECONDS_PER_DAY)))
-            .plus(UnsignedLong.valueOf(2).times(UnsignedLong.valueOf(SECONDS_PER_DAY)));
+    UnsignedLong genesis_time = eth1_timestamp;
```
Comment (cemozerr) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java]: Why did you change this? It strays from the spec: https://github.com/ethereum/eth2.0-specs/blob/v0.8.3/specs/core/0_beacon-chain.md#genesis

Comment (jrhea) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java]: good catch...i will need to look that this again.  It is potentially fallout from the confusion for the different startup modes we have.  Before that change, I was having issues with the way the slot was calculated when trying to do a sync'd start from genesis.  I saw how it was worked around before (i think it is just overwritten in another part of the code), so I might need to use a similar tactic

File: artemis/src/main/java/tech/pegasys/artemis/PeerCommand.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      footer = "Artemis is licensed under the Apache License 2.0")
+  public void generate(
+      @Mixin PeerGenerationParams params,
+      @Parameters(paramLabel = "number", description = "number of peerIDs to generate") int number)
+      throws IOException {
+    FileWriter fileWriter = new FileWriter(params.outputFile, Charset.defaultCharset());
+    PrintWriter printWriter = new PrintWriter(fileWriter);
+    printWriter.println("Private Key(Hex)\tPublic Key(Hex)\tPeerId(Base58)");
+    for (int i = 0; i < number; i++) {
+      PrivKey privKey = KeyKt.generateKeyPair(KEY_TYPE.SECP256K1).component1();
```
Comment (cemozerr) [inline artemis/src/main/java/tech/pegasys/artemis/PeerCommand.java:64]: I'm not familiar with the KeyKt libp2p module however, are you sure you're not generating the same private key over and over?

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/StartupUtil.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import tech.pegasys.artemis.datastructures.state.BeaconStateWithCache;
+import tech.pegasys.artemis.util.SSZTypes.SSZList;
+import tech.pegasys.artemis.util.SSZTypes.SSZVector;
+import tech.pegasys.artemis.util.alogger.ALogger;
+import tech.pegasys.artemis.util.alogger.ALogger.Color;
+import tech.pegasys.artemis.util.bls.BLSKeyPair;
+import tech.pegasys.artemis.util.bls.BLSSignature;
+import tech.pegasys.artemis.util.config.ArtemisConfiguration;
+
+public final class StartupUtil {
```
Comment (cemozerr) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/StartupUtil.java:45]: Nice util class. cleans things up.

File: services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                  new Object[] {this.eventBus, 0, timerPeriodInMiliseconds},
+                  new Class[] {EventBus.class, Integer.class, Integer.class});
+    } catch (IllegalArgumentException e) {
+      System.exit(1);
+    }
+  }
+
+  public void initStorage() {
+    this.chainStorageClient = ChainStorage.Create(ChainStorageClient.class, eventBus);
+    this.chainStorageClient.setGenesisTime(UnsignedLong.valueOf(config.getInteropGenesisTime()));
```
Comment (cemozerr) [inline services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java]: an if statement here to see if we're actually in interop mode might be useful. also, how are you going to set the time if we're not in interop mode?

Comment (jrhea) [reply services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java]: yes, let's talk this through

File: artemis/src/main/java/tech/pegasys/artemis/PeerCommand.java (hunk 2 above)
Comment (jrhea) [reply artemis/src/main/java/tech/pegasys/artemis/PeerCommand.java:64]: try this command:
`./gradlew run --args='peer generate 10 -o foo.dat' && cat foo.dat`

image

File: services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java (hunk 4 above)
Comment (jrhea) [reply services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java]: Also, I need the config.toml genesis time now for interop mode and when running a 4 node demo.   Thinking about renaming it

Comment (jrhea) [reply services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java]: This is the one thing i didn't fully resolve.  I think we need to rethink how our config options are grouped so it is clear what options are for 'dev' and what options are for 'prod'

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/StartupUtil.java [hunk 5]
```diff
@@ -107,18 +107,24 @@ private static Eth1Data get_eth1_data_stub(BeaconState state, UnsignedLong curre
...
+          BeaconStateUtil.initialize_beacon_state_from_eth1_new(
+              Bytes32.ZERO,
+              UnsignedLong.valueOf(config.getGenesisTime()),
+              newDeposits(config.getNumValidators()));
     }
-    return BeaconStateUtil.initialize_beacon_state_from_eth1_new(
-        Bytes32.ZERO,
-        UnsignedLong.valueOf(config.getInteropGenesisTime()),
-        newDeposits(config.getNumValidators()));
+    initialState.setGenesis_time(genesisTime);
```
Comment (cemozerr) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/StartupUtil.java:121]: It seems to me that line 121 should be in the if block instead of the else block. Because in interop, yes we do want to explicity set the time, in not interop mode, we dont want to stray any state variable from spec

**Review by cemozerr — APPROVED** [review]
Left a comment, but looks good to me other than that.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/StartupUtil.java (hunk 5 above)
Comment (jrhea) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/StartupUtil.java:121]: in non interop mode it is still a simulation.  the only real mode is through the deposit event kicking off genesis.  I tracked down those rabbit holes, but happy to walk u thru what I found

Comment (jrhea) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/StartupUtil.java:121]: in other words, I want to be able to specify the genesis time explicitly unless we are doing a run starting from deposits

Comment (cemozerr) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/StartupUtil.java:121]: Yea that makes sense, and its way cleaner than before. Good job.
